### PR TITLE
Fix: Lazy load modules depending on the task

### PR DESCRIPTION
**Pull request by molant**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [ ] ~~Added/Updated related documentation.~~
- [x] Added/Updated related tests.

## Short description of the change(s)

Use dynamic imports to load the task the user has decided to execute so
there are less modules loaded directly and the memory footprint is
smaller.

- - - - - - - - - - - - - - - - - - - - - - - - - -

Ref: https://github.com/sonarwhal/sonarwhal/issues/896#issuecomment-392114447

### CODE REVIEW

File: packages/sonarwhal/src/lib/cli/analyze.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ import * as logger from '../utils/logging';
 import { cutString } from '../utils/misc';
 import * as resourceLoader from '../utils/resource-loader';
 import { installPackages } from '../utils/npm';
-import { initSonarwhalrc } from './wizards/init';
+import initSonarwhalrc from './wizards/init';
```
Comment (sarvaje) [inline packages/sonarwhal/src/lib/cli/analyze.ts]: this should be load only if we need it.

File: packages/sonarwhal/src/lib/cli/wizards/init.ts [hunk 2]
```diff
@@ -163,8 +163,8 @@ const customConfig = async (): Promise<InitUserConfig> => {
  * * an existing published configuration package
  * * the installed resources
  */
-export const initSonarwhalrc = async (options: CLIOptions): Promise<boolean> => {
-    if (!options.init) {
+export default async (actions: CLIOptions): Promise<boolean> => {
+    if (!actions.init) {
```
Comment (sarvaje) [inline packages/sonarwhal/src/lib/cli/wizards/init.ts]: This check shouldn't be necessary. Neither in the other actions

File: packages/sonarwhal/src/lib/cli/wizards/new-parser.ts [hunk 3]
```diff
@@ -265,7 +265,7 @@ const generateFiles = async (data: NewParser) => {
  * ------------------------------------------------------------------------------
  */
 /** Generate a new parser. */
-export const newParser = async (actions: CLIOptions): Promise<boolean> => {
+export default async (actions: CLIOptions): Promise<boolean> => {
     if (!actions.newParser) {
```
Comment (sarvaje) [inline packages/sonarwhal/src/lib/cli/wizards/new-parser.ts]: This check shouldn't be necessary

File: packages/sonarwhal/src/lib/cli/actions.ts [hunk 4]
```diff
@@ -1,13 +1,29 @@
...
-import { newParser } from './wizards/new-parser';
+import { CLIOptions } from '../types';
+
+/** Wrapper to dynamically load the different CLI tasks depending on a condition */
+const action = (pkg: string, condition?: string): (actions: CLIOptions) => Promise<boolean> => {
+    return async (actions: CLIOptions): Promise<boolean> => {
+        if (!condition || actions[condition]) {
+            const { default: task } = await import(pkg);
+
+            // TODO: we shouldn't have to pass actions
```
Comment (sarvaje) [inline packages/sonarwhal/src/lib/cli/actions.ts]: I think this depends on the action. `analyze` needs it. The problem I think that is calling the parameter `actions`, it should be called `options`.

**Comment by alrra** [comment]
@molant It seems some [test hangs](https://travis-ci.org/sonarwhal/sonarwhal/jobs/383902391#L2706).

**Comment by molant** [comment]
Yes, I need to check that but it won't be until next week when I'm back from 🇪🇸
Feel free if you want to fix them (or list which ones break) 😁

**Pull request closed without merging**